fedify/federation/builder.ts

```diff
@@ -1167,6 +1191,157 @@ export class FederationBuilderImpl<TContextData>
     };
     return setters;
   }
+
+  setCollectionDispatcher<
+    TObject extends Object,
+    TParams extends Record<string, string>,
+  >(
+    name: string | symbol,
+    ...args: [
+      ConstructorWithTypeId<TObject>,
+      ParamsKeyPath<TParams>,
+      CustomCollectionDispatcher<
+        TObject,
+        TParams,
+        RequestContext<TContextData>,
+        TContextData
+      >,
+    ]
+  ): CustomCollectionCallbackSetters<
+    TParams,
+    RequestContext<TContextData>,
+    TContextData
+  > {
+    return this.#setCustomCollectionDispatcher(
+      name,
+      "collection",
+      ...args,
+    );
+  }
+
+  setOrderedCollectionDispatcher<
+    TObject extends Object,
+    TParams extends Record<string, string>,
+  >(
+    name: string | symbol,
+    ...args: [
+      ConstructorWithTypeId<TObject>,
+      ParamsKeyPath<TParams>,
+      CustomCollectionDispatcher<
+        TObject,
+        TParams,
+        RequestContext<TContextData>,
+        TContextData
+      >,
+    ]
+  ): CustomCollectionCallbackSetters<
+    TParams,
+    RequestContext<TContextData>,
+    TContextData
+  > {
+    return this.#setCustomCollectionDispatcher(
+      name,
+      "orderedCollection",
+      ...args,
+    );
+  }
+  #setCustomCollectionDispatcher<
+    TObject extends Object,
+    TParams extends Record<string, string>,
+  >(
+    name: string | symbol,
+    collectionType: "collection" | "orderedCollection",
+    itemType: ConstructorWithTypeId<TObject>,
+    path: ParamsKeyPath<TParams>,
+    dispatcher: CustomCollectionDispatcher<
+      TObject,
+      TParams,
+      RequestContext<TContextData>,
+      TContextData
+    >,
+  ): CustomCollectionCallbackSetters<
+    TParams,
+    RequestContext<TContextData>,
+    TContextData
+  > {
+    const routeName = `${collectionType}:${String(name)}`;
```

Comment:
I think you should properly split namespaces for strings and symbols (e.g., `"${collectionType}:string:${name}"`, `${collectionType}:symbol:${uniqueId(name)}`[^1]), so that these two names can be distinguished:

```typescript
"Symbol(name)"
Symbol.for("name")
```

[^1]: You should implement `uniqueId()` using a global state, e.g.:

    ```typescript
    const symbols = new WeakMap<symbol, string>();
    function uniqueId(name: symbol): string {
      let identifier = symbols.get(name);
      if (identifier == null) {
        identifier = crypto.randomUUID();
        symbols.set(name, identifier);
      }
      return identifier;
    }
    ```